src/particle_restart.cpp

```diff
@@ -0,0 +1,107 @@
+#include "openmc/particle_restart.h"
+
+#include "openmc/constants.h"
+#include "openmc/hdf5_interface.h"
+#include "openmc/mgxs_interface.h"
+#include "openmc/nuclide.h"
+#include "openmc/output.h"
+#include "openmc/particle.h"
+#include "openmc/photon.h"
+#include "openmc/random_lcg.h"
+#include "openmc/settings.h"
+#include "openmc/simulation.h"
+#include "openmc/tallies/tally.h"
+
+#include <algorithm> // for copy
+#include <array>
+#include <stdexcept>
+#include <string>
+
+namespace openmc {
+
+void read_particle_restart(Particle& p, int& previous_run_mode)
+{
+  // Write meessage
+  write_message("Loading particle restart file " +
+    settings::path_particle_restart + "...", 5);
+
+  // Open file
+  hid_t file_id = file_open(settings::path_particle_restart, 'r');
+
+  // Read data from file
+  read_dataset(file_id, "current_batch", simulation::current_batch);
```

Comment:
Use the same FILE_ATTRIB.FILETYPE standard attribute fields in reading the dataset, that were used for writing. Do not inline field names such as "generations_per_batch" ...etc.